packages/arb-token-bridge-ui/src/hooks/useTransactionHistory.ts

```diff
@@ -332,35 +337,138 @@ export const useTransactionHistory = (
     }
   )
 
   // transfers initiated by the user during the current session
   // we store it separately as there are a lot of side effects when mutating SWRInfinite
-  const { data: newTransactionsData, mutate } = useSWRImmutable<
-    MergedTransaction[]
-  >(address ? ['new_tx_list', address] : null)
+  const { data: newTransactionsData, mutate: mutateNewTransactionsData } =
+    useSWRImmutable<MergedTransaction[]>(
+      address ? ['new_tx_list', address] : null
+    )
+
+  const transactions: MergedTransaction[] = [
+    ...(newTransactionsData || []),
+    ...(txPages || [])
+  ].flat()
 
   const addPendingTransaction = useCallback(
     (tx: MergedTransaction) => {
       if (!isTxPending(tx)) {
         return
       }
 
-      mutate(currentNewTransactions => {
+      mutateNewTransactionsData(currentNewTransactions => {
         if (!currentNewTransactions) {
           return [tx]
         }
 
         return [tx, ...currentNewTransactions]
       })
     },
-    [mutate]
+    [mutateNewTransactionsData]
+  )
+
+  const updateCachedTransaction = useCallback(
+    (newTx: MergedTransaction) => {
+      // check if tx is a new transaction initiated by the user, and update it
+      const foundInNewTransactionsCache = !!newTransactionsData?.find(oldTx =>
+        isSameTransaction(oldTx, newTx)
+      )
+
+      if (foundInNewTransactionsCache) {
+        // replace the existing tx with the new tx
+        mutateNewTransactionsData(txs =>
+          txs?.map(oldTx => {
+            return { ...(isSameTransaction(oldTx, newTx) ? newTx : oldTx) }
+          })
+        )
+        return
+      }
+
+      // tx not found in the new user initiated transaction list
+      // look in the paginated historical data
+      if (!txPages) {
+        return
+      }
+
+      let pageNumberToUpdate = 0
+
+      // search cache for the tx to update
+      while (
+        !txPages[pageNumberToUpdate]?.find(oldTx =>
+          isSameTransaction(oldTx, newTx)
+        )
+      ) {
+        pageNumberToUpdate++
+
+        if (pageNumberToUpdate > txPages.length) {
+          // tx not found
+          return
+        }
+      }
+
+      const oldPageToUpdate = txPages[pageNumberToUpdate]
+
+      if (!oldPageToUpdate) {
+        return
+      }
+
+      // replace the old tx with the new tx
+      const updatedPage = oldPageToUpdate.map(oldTx => {
+        return isSameTransaction(oldTx, newTx) ? newTx : oldTx
+      })
+
+      // all old pages including the new updated page
+      const newTxPages = [
+        ...[...txPages].slice(0, pageNumberToUpdate),
+        updatedPage,
+        ...[...txPages].slice(pageNumberToUpdate + 1)
+      ]
+
+      mutateTxPages(newTxPages, false)
+    },
+    [mutateNewTransactionsData, mutateTxPages, newTransactionsData, txPages]
+  )
+
+  const updatePendingTransaction = useCallback(
+    async (tx: MergedTransaction) => {
+      const foundInCache = !!transactions.find(t => isSameTransaction(t, tx))
```

Comment:
Can we make the cache a key-value store with the key being the tx hash + chain id, for quicker retrieval?

Reply:
updated